gitprovider.go

```diff
@@ -0,0 +1,115 @@
+
+	// ErrFieldRequired specifies the case where a required field isn't populated at use time
+	ErrFieldRequired = errors.New("field is required")
+	// ErrFieldInvalid specifies the case where a field isn't populated in a valid manner
+	ErrFieldInvalid = errors.New("field is invalid")
+	// ErrFieldEnumInvalid specifies the case where the given value isn't part of the known values in the enum
+	ErrFieldEnumInvalid = errors.New("field value isn't known to this enum")
+
+	// ErrURLUnsupportedScheme is returned if an URL without the HTTPS scheme is parsed
+	ErrURLUnsupportedScheme = errors.New("unsupported URL scheme, only HTTPS supported")
+	// ErrURLUnsupportedParts is returned if an URL with fragment, query values and/or user information is parsed
+	ErrURLUnsupportedParts = errors.New("organization or repository URLs cannot have fragments, query values nor user information")
+	// ErrURLInvalid is returned if an URL is invalid when parsing
+	ErrURLInvalid = errors.New("invalid organization or repository URL")
+	// ErrURLMissingRepoName is returned if there is no repository name in the URL
+	ErrURLMissingRepoName = errors.New("missing repository name")
+)
+
+// ProviderID is a typed string for a given Git provider
+// The provider constants are defined in their respective packages
+type ProviderID string
+
+// Creator is an interface which all objects that can be created
+// (using the Client) should implement
+type Creator interface {
+	// ValidateCreate will be run in every .Create() client call, before defaulting
+	// Set (non-nil) and required fields should be validated
+	ValidateCreate() error
+	// Default will be run after validation, setting optional pointer fields to their
+	// default values before doing the POST request
+	Default()
+}
+
+// Creator is an interface which all objects that can be updated
+// (using the Client) should implement
+type Updator interface {
+	// ValidateUpdate will be run in every .Update() client call
+	// Set (non-nil) and required fields should be validated, if needed
+	// No defaulting happens for update calls
+	ValidateUpdate() error
+}
+
+// Deletor is an interface which all objects that can be deleted
+// (using the Client) should implement
+type Deletor interface {
+	// ValidateDelete will be run in every .Delete() client call
+	// Set (non-nil) and required fields should be validated, if needed
+	// No defaulting happens for delete calls
+	ValidateDelete() error
+}
+
+// Object is the interface all types should implement
+type Object interface {
+	// GetInternal returns the underlying struct that's used
+	GetInternal() interface{}
+}
+
+// InternalHolder can be embedded into other structs to implement the Object interface
+type InternalHolder struct {
+	// Internal contains the underlying object.
+	// +optional
+	Internal interface{} `json:"-"`
+}
+
+// GetInternal implements the Object interface
+func (ih InternalHolder) GetInternal() interface{} {
+	return ih.Internal
+}
+
+func boolVar(b bool) *bool {
```

Comment:
I'd keep this private but move it to `util.go` or similar (and add a one-line comment). This helps in keeping the interface definition files clean.